RelationalAI/DefaultAccessTokenHandler.cs

```
@@ -0,0 +1,106 @@
+
+            token = await rest.RequestAccessTokenAsync(host, creds);
+            if (token != null)
+            {
+                WriteAccessToken(creds, token);
+            }
+
+            return token;
+        }
+
+        private string CacheName()
+        {
+            var envHome = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "HOMEPATH" : "HOME";
+            var home = Environment.GetEnvironmentVariable(envHome);
+
+            return Path.Join(home, ".rai", "tokens.json");
+        }
+
+        private AccessToken ReadAccessToken(ClientCredentials creds)
+        {
+            var cache = ReadTokenCache();
+            if (cache == null)
+            {
+                return null;
+            }
+
+            if (cache.ContainsKey(creds.ClientId))
+            {
+                return cache[creds.ClientId];
+            }
+
+            return null;
+        }
+
+        private Dictionary<string, AccessToken> ReadTokenCache()
+        {
+            try
+            {
+                var data = File.ReadAllText(CacheName());
+                var cache = JsonConvert.DeserializeObject<Dictionary<string, AccessToken>>(data);
+                return cache;
+            }
+            catch (IOException)
+            {
+            }
+
+            return null;
+        }
+
+        private void WriteAccessToken(ClientCredentials creds, AccessToken token)
+        {
+            try
+            {
+                var cache = ReadTokenCache();
+                var dict = cache ?? new Dictionary<string, AccessToken>();
+                if (dict.ContainsKey(creds.ClientId))
+                {
+                    dict[creds.ClientId] = token;
+                }
+                else
+                {
+                    dict.Add(creds.ClientId, token);
+                }
+
+                File.WriteAllText(CacheName(), JsonConvert.SerializeObject(dict));
+            }
+            catch (IOException)
+            {
+            }
```

yeah, I don't feel very comfortable with swallowing these errors. What's the rationale behind that?

---

if for any reason we can't get the token cache, the sdk should fall back to requesting a new access token without blocking the sdk users

---

it's fine to fall back to requesting new access token but was wondering if we should at least notify the user in case of broken expectations. Do we use logging at all in the sdks?

---

No we don't use logging currently, I'll add a todo there since I'll be working on adding on logging framework and this is a potential function where we need to log a warning or a debug message